pkg/receive/handler.go

```diff
@@ -364,42 +372,31 @@ func (h *Handler) getStats(r *http.Request, statsByLabelName string) ([]statusap
 
 // Close stops the Handler.
 func (h *Handler) Close() {
-	if h.listener != nil {
-		runutil.CloseWithLogOnErr(h.logger, h.listener, "receive HTTP listener")
-	}
+	runutil.CloseWithLogOnErr(h.logger, h.httpSrv, "receive HTTP server")
```

Comment:
Now thinking about it - what if `Close()` is called before `h.httpSrv` is created? I think there's a race here too which warrants a lock. See here: https://github.com/oklog/run/blob/main/group.go#L42-L48.

To trigger the race, it should be enough to error out here between `Run()` finishes and `Close()`: https://github.com/thanos-io/thanos/blob/main/cmd/thanos/receive.go#L373-L375 in this actor routine and any of the above it.

Reply:
`Close cannot be run before `h.httpSrv` is created since it is created by `NewHandler()` and the factory is called before the group actor is added: https://github.com/thanos-io/thanos/blob/main/cmd/thanos/receive.go#L390

Reply:
Ah, sorry, somehow looked over the part above where you've moved setting `h.httpSrv` into the constructor!